Add slot swapping to the hotbar via shift click

Emirlol opened the pull request:
Also adds kotlin, ehe
The main functionality is already in the initial commit but there's some stuff to do
- [x] Save the mappings to config
- [x] Make the colors configurable
- [x] Maybe enforce 1-1 mapping of the slots
- [x] Clean up code, it's really messy as I went through a few different ways of implementing without cleaning up afterwards and it's just all around the place

AzureAaron commented:
Will fix #680

justino901 commented:
Not a dev myself so i don't know how plausible this is but i think being able to slot swap to your helmet would be nice as well for Bonzo head swapping in the middle of a dungeon. you can ignore this if it would be a pain to implement though

justino901 commented:
so i just found a visual glitch with the slot swaping. when opening the recipe book the overlay stays in position and doesnt move with inventory ui to the side

Emirlol commented:
> Not a dev myself so i don't know how plausible this is but i think being able to slot swap to your helmet would be nice as well for Bonzo head swapping in the middle of a dungeon. you can ignore this if it would be a pain to implement though

This is not possible unless you put the armor in your hotbar. I will make mapping from hotbar to the armor slots possible for this reason, in case you want to do that.

Emirlol commented:
> so i just found a visual glitch with the slot swaping. when opening the recipe book the overlay stays in position and doesnt move with inventory ui to the side

Fixed by [`5fb67c3`](https://github.com/SkyblockerMod/Skyblocker/pull/864/commits/5fb67c364dd1a0f87294a8614a436803af3cdc39). Seems like any UI movement or size change breaks it since the x and y values change, so I changed it to calculate the x and y values every frame rather than caching them for all eternity.

justino901 commented:
everything look good now havent found any bugs or problems since updating to the newest build with this. just waiting for it to be reviewed and merged into master i guess

kevinthegreat1 commented on src/main/kotlin/de/hysky/skyblocker/skyblock/SlotSwap.kt:
```diff
@@ -0,0 +1,182 @@
+		renderRectanglesAndLine(drawContext, slotAtKeyPress!!, currentSlot, CONFIGURING_SOURCE_SLOT_COLOR, CONFIGURING_TARGET_SLOT_COLOR)
+	}
+
+	private fun InventoryScreen.getFocusedSlotId() = (this as HandledScreenAccessor).focusedSlot?.let { slot -> if (isSlotValid(slot.id)) slot.id else null }
```
How does this work again? Does this pollute `InventoryScreen` so that `getFocusedSlotId` is availbale to everyone or just this file?

kevinthegreat1 replied:
Oh looks like you need to import it.

Emirlol replied:
> How does this work again? Does this pollute `InventoryScreen` so that `getFocusedSlotId` is availbale to everyone or just this file?

It should be only available within this file since it's private. From what I know, it takes the class it is extending as an argument behind the scenes, so it's essentially a static util method. You static import it like in java.
Note that extension functions can also be added in normal classes (this SlotSwap file is a singleton `object`, so everything in it is essentially static), in that case it's just a normal method and can't be used from outside the class. The method itself has to be static to be used from other files.

kevinthegreat1 reviewed (changes requested):
Only thing to do is make `util` package and class names ending in `Util` plural. But also please respond to the comments.  

Thanks for cleaning up RenderHelper, but at least #849 is not going to be happy.

Emirlol commented on src/main/kotlin/de/hysky/skyblocker/skyblock/SlotSwap.kt:
```diff
@@ -0,0 +1,182 @@
+	//This field is used instead of the field in MinecraftClient, so we don't have to type check the current screen every time.
+	private var currentScreen: InventoryScreen? = null
+	private var slotAtKeyPress: Int? = null
+	private val configureKeybinding: KeyBinding = KeyBindingHelper.registerKeyBinding(KeyBinding("key.skyblocker.slotSwapConfigure", GLFW.GLFW_KEY_L, "key.categories.skyblocker"))
+	private val resetKeybinding: KeyBinding = KeyBindingHelper.registerKeyBinding(KeyBinding("key.skyblocker.slotSwapReset", GLFW.GLFW_KEY_UNKNOWN, "key.categories.skyblocker"))
```
You mean to say that the shift-click should be configurable? I'd like to keep shift-click as default, so people who came from 1.8 with NEU can act upon their muscle memories. With that in mind, I don't know how I could make combination keybinds configurable. One of these considerations has to go unless someone knows how to do that.

Emirlol commented on src/main/kotlin/de/hysky/skyblocker/skyblock/SlotSwap.kt:
```diff
@@ -0,0 +1,182 @@
+			configureKeybinding.matchesKey(key, scancode) -> {
+				val slotAtKeyRelease = screen.getFocusedSlotId()
+				if (slotAtKeyPress != null && slotAtKeyRelease != null && slotAtKeyPress != slotAtKeyRelease
+					&& (isHotbarSlot(slotAtKeyPress!!) || isHotbarSlot(slotAtKeyRelease))
```
It can be `null` if the key press happened while the cursor wasn't on a proper slot as you can see from `InventoryScreen.getFocusedSlotId()`. If `focusedSlot` is `null,` the whole expression evaluates to `null`. If it's not `null`, and then the slot is not valid for slot swap (not a part of inventory of armor slots), it's still going to end up as `null`.

Emirlol commented:
> Only thing to do is make `util` package and class names ending in `Util` plural.

Done in 35b633c. I find it quite interesting that I didn't notice the plural style we had for so long.
> Thanks for cleaning up RenderHelper, but at least #849 is not going to be happy.

Should be good now since it's merged to master and this is rebased onto main.

AzureAaron reviewed (changes requested):
This PR needs some core changes
- Don't depend on or use FLK, manually JiJ any Kotlin stuff needed (really only the standard library I see as being useful)
- Coroutines are unnecessary as Java has its own concurrency stuff, the current usage can and probably should use the `Scheduler` unless the operation is blocking at which point CompletableFutures should be used
- No separate source sets, Kotlin files must go into the Java source set - NEUs codebase is difficult to navigate because of this and we will have only very few Kotlin files anyways

No clue if Kotlin can work in Eclipse so I will need to investigate that later

Emirlol commented:
I'm kinda lazy to clean this up so I've given up on kotlin for skyblocker. I've copied everything to a separate mod instead, which can be found [here](https://github.com/Emirlol/ItemSwap) for anyone interested.
